parlai/core/script.py

```diff
@@ -0,0 +1,266 @@
+from parlai.core.loader import register_script, SCRIPT_REGISTRY  # noqa: F401
+
+
+def setup_script_registry():
+    """
+    Loads the scripts so that @register_script is hit for all.
+    """
+    for module in pkgutil.iter_modules(parlai.scripts.__path__, 'parlai.scripts.'):
+        importlib.import_module(module.name)
+
+
+class ParlaiScript(object):
+    """
+    A ParlAI script is a standardized form of access.
+    """
+
+    parser: ParlaiParser
+
+    @classmethod
+    @abstractmethod
+    def setup_args(cls) -> ParlaiParser:
+        """
+        Create the parser with args.
+        """
+        # we want to later deprecate this for add_cmdline_args
+        pass
+
+    def __init__(self, opt: Opt):
+        self.opt = opt
+
+    @abstractmethod
+    def run(self):
+        """
+        The main method.
+
+        Must be implemented by the script writer.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def _run_kwargs(cls, kwargs: Dict[str, Any]):
+        """
+        Construct and run the script using kwargs, pseudo-parsing them.
+        """
+        parser = cls.setup_args()
+        opt = parser.parse_kwargs(**kwargs)
+        return cls._run_from_parser_and_opt(opt, parser)
+
+    @classmethod
+    def _run_args(cls, args: Optional[List[str]] = None):
+        """
+        Construct and run the script using args, defaulting to getting from CLI.
+        """
+        parser = cls.setup_args()
+        opt = parser.parse_args(args=args, print_args=False)
+        return cls._run_from_parser_and_opt(opt, parser)
+
+    @classmethod
+    def _run_from_parser_and_opt(cls, opt: Opt, parser: ParlaiParser):
+        script = cls(opt)
+        script.parser = parser
+        return script.run()
+
+    @classmethod
+    def main(cls, *args, **kwargs):
+        """
+        Run the program, possibly with some given args.
+
+        You may provide command line args in the form of strings, or
+        options. For example:
+
+        >>> MyScript.main(['--task', 'convai2'])
+        >>> MyScript.main(task='convai2')
+
+        You may not combine both args and kwargs.
+        """
+        assert not (bool(args) and bool(kwargs))
+        if args:
+            return cls._run_args(args)
+        elif kwargs:
+            return cls._run_kwargs(kwargs)
+        else:
+            return cls._run_args(None)
+
+    @classmethod
+    def help(cls, **kwargs):
+        f = io.StringIO()
+        parser = cls.setup_args()
+        parser.prog = cls.__name__
+        parser.add_extra_args(parser._kwargs_to_str_args(**kwargs))
+        parser.print_help(f)
+        return f.getvalue()
+
+
+class _SupercommandParser(ParlaiParser):
+    """
+    Specialty ParlAI parser used for the supercommand.
+
+    Contains some special behavior.
+    """
+
+    def __init__(self, *args, **kwargs):
+        from parlai.utils.strings import colorize
+
+        logo = ""
+        logo += colorize('       _', 'red') + "\n"
```

Comment:
lol

Reply:
this is so extra

Reply:
![obama](https://media.giphy.com/media/3oJtgiUAQcwNaQrs4M/giphy.gif)

Reply:
seriously, gotta procrastinate on PSC one way or another.